### PR TITLE
add heltec_v4 board.

### PR DESCRIPTION
**Pull request by Quency-D**
Features of heltec_v4:
1.Equipped with ESP32-S3R2 chip, featuring built-in 2MB PSRAM and external 16MB Flash, making it more suitable for UI and complex system design;
2.SX-1262 LoRa chip, high-power version with LoRa transmission power increased to 27±1dBm;
3.Form factor and pin compatibility with WiFi LoRa 32 V2 and V3;
4.PC casing protects the screen and integrates a 2.4G antenna, enhancing overall durability and integration;  
5.Supports Wi-Fi b/g/n, Bluetooth(BLE), and LoRa communication, with a reserved IPEX (U.FL) interface;
6.Supports solar power: Added SH1.25-2P solar interface, compatible with 4.4-6V/540mA solar panel input;  
7.Optimized lithium battery management: Supports charge/discharge management, overcharge protection, power level detection, and automatic 	switching between USB and battery power, with compatibility for solar charging;
8.Added SH1.25-8Pin GPS interface for easy expansion of positioning functionality.  
9.Type-C USB interface with integrated voltage regulation, ESD protection, short-circuit protection, and RF isolation design;

## 🤝 Attestations

- [x] I have tested that my proposed changes behave as described.
- [x] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [x] Heltec (Lora32) V3
  - [x] Heltec (Lora32) V4

### CODE REVIEW

**Comment by gigawatts** [comment]
Compile failed for me until I added a new entry in `src/mesh/generated/meshtastic/mesh.pb.h` after line 274, otherwise it is upset that `meshtastic_HardwareModel_HELTEC_V4` is undefined in `src/platform/esp32/architecture.h` line 201.

https://github.com/Quency-D/firmware/pull/2

Flashed to my V4, working so far.

**Comment by Quency-D** [comment]
> Compile failed for me until I added a new entry in `src/mesh/generated/meshtastic/mesh.pb.h` after line 274, otherwise it is upset that `meshtastic_HardwareModel_HELTEC_V4` is undefined in `src/platform/esp32/architecture.h` line 201.
> 
> [Quency-D#2](https://github.com/Quency-D/firmware/pull/2)
> 
> Flashed to my V4, working so far.

This place should be automatically generated by protobuf later.

**Comment by gigawatts** [comment]
> > Compile failed for me until I added a new entry in `src/mesh/generated/meshtastic/mesh.pb.h` after line 274, otherwise it is upset that `meshtastic_HardwareModel_HELTEC_V4` is undefined in `src/platform/esp32/architecture.h` line 201.
> > [Quency-D#2](https://github.com/Quency-D/firmware/pull/2)
> > Flashed to my V4, working so far.
> 
> This place should be automatically generated by protobuf later.

Later in the compile? I must be missing some context / understanding of the build process. I followed [this process](https://meshtastic.org/docs/development/firmware/build/) from your fork, but it failed to compile until I defined that variable.

**Comment by sanderDijkxhoorn** [comment]
Hi there, thanks for the work.

Building failed for me too, however after running ./bin/regen-protos.sh it build and flashed fine, thanks!

File: variants/esp32s3/heltec_v4/platformio.ini [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[env:heltec-v4] 
+extends = esp32s3_base
+board = heltec_v4
+board_level = pr
```
Comment (vidplace7) [inline variants/esp32s3/heltec_v4/platformio.ini]:
```suggestion
```
We probably don't want to build this upon *every* PR. There are plenty of esp32s3 test targets.

**Comment by karman85** [comment]
Hello, 

I would like to add the alpha schematic:

https://cdn.discordapp.com/attachments/1351808282813136947/1415260721549279274/WiFi_LoRa_32_V4.1_alpha.pdf

and the GC1109 datasheet:

https://www.geochipinc.com/uploads/23940615_1756826539.pdf

The GC1109 states a maximum 20dBm TX_IN when working in bypass, and 5dBm TX_IN when doing amplification. The SX1262 can outpout 22dBm max, and the aplha schematic shows a 15dB pi atenuator. This would still give 7dBm in TX_IN. 

As I understand, Meshtastic firmware should limit SX1262 RF output to 20dBm max to not burn the PA in the GC1109.

This would be fixed by adding TX_GAIN_LORA = 7 and SX126X_MAX_POWER = 20 to the variant.

Also the best way would be defining a new radio in the src/configuration.h file with this two parameters.

Regards

**Comment by Richard3366** [comment]
> Hello,
> 
> I would like to add the alpha schematic:
> 
> https://cdn.discordapp.com/attachments/1351808282813136947/1415260721549279274/WiFi_LoRa_32_V4.1_alpha.pdf
> 
> and the GC1109 datasheet:
> 
> https://www.geochipinc.com/uploads/23940615_1756826539.pdf
> 
> The GC1109 states a maximum 20dBm TX_IN when working in bypass, and 5dBm TX_IN when doing amplification. The SX1262 can outpout 22dBm max, and the aplha schematic shows a 15dB pi atenuator. This would still give 7dBm in TX_IN.
> 
> As I understand, Meshtastic firmware should limit SX1262 RF output to 20dBm max to not burn the PA in the GC1109.
> 
> This would be fixed by adding TX_GAIN_LORA = 25 and SX126X_MAX_POWER = 5 to the variant.
> 
> Also the best way would be defining a new radio in the src/configuration.h file with this two parameters.
> 
> Regards

This datasheet and schematic are only applicable to the current batch of several dozen test boards and may differ from the officially released version.

It has been confirmed that the official version will use a 17dB attenuator.

**Comment by karman85** [comment]
> > Hello,
> > I would like to add the alpha schematic:
> > https://cdn.discordapp.com/attachments/1351808282813136947/1415260721549279274/WiFi_LoRa_32_V4.1_alpha.pdf
> > and the GC1109 datasheet:
> > https://www.geochipinc.com/uploads/23940615_1756826539.pdf
> > The GC1109 states a maximum 20dBm TX_IN when working in bypass, and 5dBm TX_IN when doing amplification. The SX1262 can outpout 22dBm max, and the aplha schematic shows a 15dB pi atenuator. This would still give 7dBm in TX_IN.
> > As I understand, Meshtastic firmware should limit SX1262 RF output to 20dBm max to not burn the PA in the GC1109.
> > This would be fixed by adding TX_GAIN_LORA = 25 and SX126X_MAX_POWER = 5 to the variant.
> > Also the best way would be defining a new radio in the src/configuration.h file with this two parameters.
> > Regards
> 
> This datasheet and schematic are only applicable to the current batch of several dozen test boards and may differ from the officially released version.
> 
> It has been confirmed that the official version will use a 17dB attenuator.

Yes. I didnt saw the atenuator (was writen in chinese and there aren't values writen down).

So current documentation and alpha boards should use:

TX_GAIN_LORA = 7
SX126X_MAX_POWER = 20

as per written confirmation (still no docs nor hardware of the official release) future boards should be fine with the 17dB atenuator and:

TX_GAIN_LORA = 5
SX126X_MAX_POWER = 22

**Comment by fifieldt** [comment]
Conveniently, SX126X_MAX_POWER = 22 is the default, so current settings should be fine for official release.

Let's set TX_GAIN_LORA = 5 in this patch's variant.h

File: variants/esp32s3/heltec_v4/variant.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#define SX126X_BUSY LORA_DIO2
+#define SX126X_RESET LORA_RESET
+
+#define SX126X_DIO2_AS_RF_SWITCH
+#define SX126X_DIO3_TCXO_VOLTAGE 1.8
+
+#define LORA_PA_POWER  7  //power en
+#define LORA_PA_EN     2  
+#define LORA_PA_TX_EN  46  //enable tx
+
```
Comment (fifieldt) [inline variants/esp32s3/heltec_v4/variant.h:43]: @Quency-D , we think we should set
#define TX_GAIN_LORA = 5
here to deal with the power amp + attenuation settings that will be used for official board.

Comment (Quency-D) [reply variants/esp32s3/heltec_v4/variant.h:43]: I don't understand why TX_GAIN_LORA needs to be configured, because the subsequent official version will not damage the amplifier chip after attenuation.

Comment (fifieldt) [reply variants/esp32s3/heltec_v4/variant.h:43]: We use this value to keep within the regulatory limits for the frequencies. Without it, the device may not be legal to use in some of the markets.

Comment (fifieldt) [reply variants/esp32s3/heltec_v4/variant.h:43]: Math might be easier ....

eg
SX126X_MAX_POWER = 22 
TX_GAIN_LORA = PA gain - attenuator loss = 20 - 17  =  3


This must be true to be legal:
SX126X_MAX_POWER + TX_GAIN_LORA < region power limit

So, for a region like USA 902 ( 30dBm limit)

22 + 3 < 30 

so we can use SX126X_MAX_POWER (22)

For a region like Singapore 923 (20 dBm limit):

22 + 3 >> 20 

So we have to use the region power limit (20)

Comment (Quency-D) [reply variants/esp32s3/heltec_v4/variant.h:43]: OK, I understand what you mean. However, this amplifier chip is not linear. We will arrange testing as soon as possible to ensure its maximum output meets the regulations.

Comment (fifieldt) [reply variants/esp32s3/heltec_v4/variant.h:43]: Thats great! Thank you very much!

Comment (gigawatts) [reply variants/esp32s3/heltec_v4/variant.h:43]: So for the use of the alpha boards, dump these into `heltec_v4/variant.h` ?
```
#define TX_GAIN_LORA 7
#define SX126X_MAX_POWER 20
```

Comment (fifieldt) [reply variants/esp32s3/heltec_v4/variant.h:43]: @Quency-D , would it be OK to insert some "conservative" values first so we can merge this PR, then update the values later once testing is complete?

It's your hardware, so whatever way you want to do it is fine. I was just thinking I  merging this patch faster will help community testers get firmware more easily.

**Comment by fifieldt** [comment]
Tested this and it flashes well on the heltec v4 test board :)

File: variants/esp32s3/heltec_v4/variant.h (hunk 2 above)
Comment (fifieldt) [reply variants/esp32s3/heltec_v4/variant.h:43]: Oh, and also, if you do want us to implement a non-linear algorithm, we can probably do that :) Similar to battery OCV_ARRAY. We know how important power is for the users!

Comment (Quency-D) [reply variants/esp32s3/heltec_v4/variant.h:43]: > @Quency-D , would it be OK to insert some "conservative" values first so we can merge this PR, then update the values later once testing is complete?
> 
> It's your hardware, so whatever way you want to do it is fine. I was just thinking I merging this patch faster will help community testers get firmware more easily.

I agree with this idea. By setting the maximum output of SX1262 to 11, the overall output can be limited to 22.

**Review by fifieldt — APPROVED** [review]

**Comment by gigawatts** [comment]
Unsure if this is limited to the v4 hardware, or is a broader issue with the version this was branched from (2.7.8), but Bluetooth seems to lockup after a few days of uptime. All the other main functions (LoRa radio, local display UI, etc) keep working. The `System` tab on the local display shows Heap at  `! 99%`

Update: never mind --> https://github.com/meshtastic/firmware/releases/tag/v2.7.8.a0c0388
